frame/im-online/src/lib.rs

```diff
@@ -103,6 +103,10 @@ use frame_system::offchain::{
 	SendTransactionTypes,
 	SubmitTransaction,
 };
+use pallet_session::{
```

Comment:
As I've mentioned earlier - the only decoupling that would be acceptable for me is if we get rid of `pallet_session` direct dependency.

I know this is now only required for offences reporting, but this also clearly proves that the decoupling is incomplete.
What happens if I report an offence to the `session` module directly, but the validator is simply not there, cause it was returned by `SessionInterface` abstraction.
We need to make sure that `SessionInterface` encompasses all the things related to validator management - i.e. if you want to return a custom list of validators which doesn't come from `session` you should as well handle offences of these validators - hence this two things should be couple into a single interface (`trait`).
I.e. the abstraction should really minimize making a stupid mistake when implementing it.

Reply:
>What happens if I report an offence to the session module directly, but the validator is simply not there, cause it was returned by SessionInterface abstraction.

hmm, how can you report an offence to `session` directly? I don't see how this happens.

Your concern is pretty right and I have spotted one problem in current code that `Keys` in `im-online` are handled inconsistently. We do need a more complete design for the initial purpose of this PR. Just let me know if you are still not much clear about the intention of it.

Reply:
> hmm, how can you report an offence to session directly? I don't see how this happens.

It was a simplification, since we report the offences using `ReportOffence` trait, but I meant the fact that it uses the types from `pallet_session` (`FullIdentification`).